src/Illuminate/Container/Container.php

```diff
@@ -234,6 +234,10 @@ public function bind($abstract, $concrete = null, $shared = false)
         // bound into this container to the abstract type and we will just wrap it
         // up inside its own Closure to give us more convenience when extending.
         if (! $concrete instanceof Closure) {
+            if (! is_string($concrete)) {
+                throw new \TypeError(self::class.'::bind(): Argument #2 ($concrete) must be of type Closure|string|null');
```

Comment:
Note for Taylor: This error message format is what PHP 8.0 does when there's a genuine union type.